### DISCUSSION TITLE
Use KEDA for DNS Horizontal Autoscaling

### DISCUSSION THREAD

**Discussion by alekhrycaiko** [kedacore/keda#4318] [Q&A / Need Help]
Hi, I've been reading the docs on [horizontal autoscaling for DNS services ](https://kubernetes.io/docs/tasks/administer-cluster/dns-horizontal-autoscaling/). This recommends using the [cluster-proportional-autoscaler](https://github.com/kubernetes-sigs/cluster-proportional-autoscaler) for autoscaling DNS services. However, I'd like to use KEDA since that's already installed on my system. 

I haven't seen any documentation about the best way to do this and was curious about how the community suggests to do this.

**Comment by tomkerkhove** [comment]
Interesting use-case - Can you elaborate on the scenario more please as we haven't had this request before

**Comment by alekhrycaiko** [comment]
Hi @tomkerkhove 

The example use case is that I'd like to scale a DNS based service such as CoreDNS. CoreDNS can run into a few failure modes that can be alleviated by adding Pods in relation to the cluster's Node or CPU count. Hence the `cluster-proportional-autoscaler` is recommended. 

However, to do the scaling mentioned above in KEDA appears difficult right now. I think this is largely because I'd be using the Node or CPU count as a threshold to breach. Node and CPU counts in growing systems don't generally decrease often. Due to that, the KEDA model right now seems to lead to continuous up scaling.

Ideally I'd like to scale between my min/max based on some kind of ratio. For example, for every 100 CPU add 1 replica. The linear scaling of `cluster-proportional-autoscaler` is a good example of this. I don't currently see a way to get this kind of functionality from KEDA but if I'm missing something I'd definitely be receptive to suggestions 🙇

**Reply by zroubalik** [reply]
How do you get those metrics? If it is something that can be scraped by Prometheus, then you are just a query from the desired behavior :)

**Reply by alekhrycaiko** [reply]
Hi @zroubalik yes you could use Prometheus to get node count and pod count. I'm less clear on how you'd actually define a query and threshold here similar to how I'm describing however.

A couple examples of things I've evaluated...

if you were to use `count(kube_node_info)` and set it  to 20, a continuous up scale would occur once you breach that value. From a maintainer perspective that isn't great as every time we have a scaling event (cluster size grows), I have to manage the max and change the threshold. Some kind of proportional scaling would be better. 

A slightly better case I thought of was using a proportion via Prometheus, but again, there's some issues. 

For example, `1-max(kube_deployment_spec_replicas{deployment="coredns") / count(kube_node_info)` can provide a proportion of nodes lacking my CoreDNS pod. This doesn't really hold up if my deployment count exceeds my node count though. 

I haven't come up with a good way to do the equivalent of for 'every 100 nodes add a replica' in Prometheus. But if you have a suggestion there, let me know 👍
